crates/egui/src/widgets/image.rs

```diff
@@ -769,3 +785,39 @@ pub fn paint_texture_at(
         }
     }
 }
+
+fn get_index(ctx: &Context, uri: &str) -> usize {
+    let now = ctx.input(|i| Duration::from_secs_f64(i.time));
+
+    let durations: Option<Arc<Vec<Duration>>> =
+        ctx.data(|data| data.get_temp(ImageDataIdIndex.id(uri)));
+    if let Some(durations) = durations {
+        let frames: Duration = durations.iter().sum();
+        let pos = now.as_millis() % frames.as_millis().max(1);
+        let mut cumulative_duration = 0;
+        let mut index = 0;
+        for (i, duration) in durations.iter().enumerate() {
+            cumulative_duration += duration.as_millis();
+            if cumulative_duration >= pos {
+                index = i;
+                break;
+            }
+        }
+        if let Some(duration) = durations.get(index) {
+            ctx.request_repaint_after(*duration);
+        }
+        index
+    } else {
+        0
+    }
+}
+
+#[derive(Clone, Copy, PartialEq, Eq, Hash)]
+pub struct ImageDataIdIndex;
```

Comment:
I suggest this:


```suggestion
/// The number of frames in a GIF, and their individual durations.
#[derive(Clone)]
pub struct GifFrameDurations(pub Arc<Vec<Duration>>);
```

and then you put that into the data store, meaning you know that you put the exact correct type into the data store that you later pull out.